frameworks/keyed/sauron/src/lib.rs

```diff
@@ -0,0 +1,236 @@
+use rand::prelude::*;
+use sauron::prelude::*;
+use std::cmp::min;
+use web_sys::window;
+
+static ADJECTIVES: &[&str] = &[
+    "pretty",
+    "large",
+    "big",
+    "small",
+    "tall",
+    "short",
+    "long",
+    "handsome",
+    "plain",
+    "quaint",
+    "clean",
+    "elegant",
+    "easy",
+    "angry",
+    "crazy",
+    "helpful",
+    "mushy",
+    "odd",
+    "unsightly",
+    "adorable",
+    "important",
+    "inexpensive",
+    "cheap",
+    "expensive",
+    "fancy",
+];
+
+static COLOURS: &[&str] = &[
+    "red", "yellow", "blue", "green", "pink", "brown", "purple", "brown", "white", "black",
+    "orange",
+];
+
+static NOUNS: &[&str] = &[
+    "table", "chair", "house", "bbq", "desk", "car", "pony", "cookie", "sandwich", "burger",
+    "pizza", "mouse", "keyboard",
+];
+
+#[derive(Clone, PartialEq)]
+struct RowData {
+    id: usize,
+    label: String,
+}
+
+impl RowData {
+    fn new(id: usize, rng: &mut SmallRng) -> Self {
+        let adjective = *ADJECTIVES.choose(rng).unwrap();
+        let colour = *COLOURS.choose(rng).unwrap();
+        let noun = *NOUNS.choose(rng).unwrap();
+
+        let label = [adjective, colour, noun].join(" ");
+
+        Self { id, label }
+    }
+    fn view(&self, selected: bool) -> Node<Msg> {
+        let id = self.id;
+        node! {
+            <tr class={if selected { "danger" } else  { "" }}>
```

Comment:
I noticed that the `key` attribute is not set in the code.
If there is no key specified, sauron will aggressively reuse similar nodes.
If there is a key specified, if the key differ the node will be always discarded.

I still have a hard time running the js-framework-benchmark myself, but I would say adding a key here would most likely solve the issue.

```html
            <tr class={if selected { "danger" } else  { "" }} key=id >
```

Reply:
Please let me know if I can help running the benchmark on your machine. Where did you get stucked?

Reply:
thanks, it looks like this was the trick ... but I also updated sauron to `0.50.1`.. so I don't know what causes it ;-)

Reply:
> Please let me know if I can help running the benchmark on your machine. Where did you get stucked?

I chrome is not installed on a usual place on my machine (NixOS) so this is the error:

```
Executing frameworks/keyed/sauron and benchmark 01_run1k failed: browserType.launch: Failed to launch chromium because executable doesn't exist at /usr/bin/google-chrome
```
So how can I set the path to a custom location?

Reply:
The update to 0.50.1 should have no effect to the code. 
> `Executing frameworks/keyed/sauron and benchmark 01_run1k failed: browserType.launch: Failed to launch chromium because executable doesn't exist at /usr/bin/google-chrome`

It looks like you need to have google-chrome installed

Reply:
There's an option for that 😄 :
`npm run bench -- --chromeBinary /bla/bla keyed/vanillajs` and `npm run isKeyed -- --chromeBinary /bla/bla keyed/vanillajs` (in directory webdriver-ts) should both work with custom chrome locations. Can you try that?

Reply:
@krausest thanks a lot, now my CPU is sweating :laughing:

Reply:
@ivanceras 

The `isKeyed` test fails with this:
```
Keyed test for swap failed. Swap must add the TRs that it removed, but there were 997 new nodes
sauron-v0.50.1-keyed is keyed for 'run benchmark' and keyed for 'remove row benchmark' and non-keyed for 'swap rows benchmark'
```

Reply:
@flosse  I'm not sure what `isKeyed` test does? but what should be the expected behaviour?

Reply:
I'm able to run the benchmark now, I'm taking a look.

Reply:
The isKeyed test checks whether the implementation behaves keyed. As such replacing rows must create new dom nodes (and not reuse existing ones), removing a row must remove the row you clicked on (and not e.g. the last row) and swap must actually move the dom nodes (and not create new nodes). 

Seems like swapping rows in sauron added 997 new tr rows. You can read here a bit more about that: https://github.com/krausest/js-framework-benchmark/pull/1060#issuecomment-1168247794

Reply:
keyed diffing in sauron is only designed from top to bottom matching, that's why this fails. It may take me awhile to try to include heuristic to start from bottom to top in the existing code.
Meanwhile @flosse, maybe we can move sauron to the non-keyed frameworks for now.

Reply:
> Meanwhile @flosse, maybe we can move sauron to the non-keyed frameworks for now.

done!